sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py

```diff
@@ -1217,23 +1215,33 @@ def get_ancestors(name):
 
 
 def _parent_for_fused_stages(stages, context):
-  # type: (Iterable[Optional[str]], TransformContext) -> Optional[str]
+  # type: (Iterable[Stage], TransformContext) -> Optional[str]
 
   '''Returns the name of the new parent for the fused stages.
 
   The new parent is the lowest common ancestor of the fused stages that is not
   contained in the set of stages to be fused. The provided context is used to
   compute ancestors of stages.
   '''
+
+  parents = context.parents_map()
+  # If any of the input stages were produced by fusion or an optimizer phase,
```

Comment:
Is there a case where the parent's parent is not correctly set?

Reply:
Nope-ish. This problem really arises because for stages going through the phases, there are two sources of truth for their parents: `parent` field of the stage and the `subtransforms` field of transforms in the `PipelineContext`. The parent of a stage _has_ to be a transform in PipelineContext, which means there is only one source of truth (the `subtransforms` field). So we don't encounter this issue with grandparents.

Reply:
Ack. Thanks for the clarification.